ARROW-13404: [Doc][Python] Improve PyArrow documentation for new users (#10999)
apache/arrow: docs/source/python/getstarted.rst

.. Licensed to the Apache Software Foundation (ASF) under one
.. or more contributor license agreements.  See the NOTICE file
.. distributed with this work for additional information
.. regarding copyright ownership.  The ASF licenses this file
.. to you under the Apache License, Version 2.0 (the
.. "License"); you may not use this file except in compliance
.. with the License.  You may obtain a copy of the License at

..   http://www.apache.org/licenses/LICENSE-2.0

.. Unless required by applicable law or agreed to in writing,
.. software distributed under the License is distributed on an
.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
.. KIND, either express or implied.  See the License for the
.. specific language governing permissions and limitations
.. under the License.

.. _getstarted:

Getting Started
===============

> Arrow manages data in Arrays (:class:`pyarrow.Array`), which can be

Reviewer: Lowercase "Arrays" since "tables" is lowercased just after?

Author: :+1:

grouped in tables (:class:`pyarrow.Table`) to represent columns of data
in tabular data.

Arrow also provides support for various formats to get those tabular
data in and out of disk and networks. Most commonly used formats are

> Parquet (:ref:`parquet`) and the IPC format (:ref:`ipc`). 

Reviewer: I'm not sure the IPC format is really commonly used compared to, say, CSV :-)

Author: Here I mentioned the formats that are column major. As CSV and JSON are row oriented I didn't mentioned them as primary choices, but they are mentioned in the "Saving and Loading Tables" section together with the other available formats.

Once you have tabular data, Arrow provides out of the box
the features to save and restore that data for common formats
like Parquet:

.. ipython:: python   

    import pyarrow.parquet as pq

    pq.write_table(birthdays_table, 'birthdays.parquet')

Once you have your data on disk, loading it back is a single function call,
and Arrow is heavily optimized for memory and speed so loading
data will be as quick as possible

.. ipython:: python

    reloaded_birthdays = pq.read_table('birthdays.parquet')

    reloaded_birthdays

Saving and loading back data in arrow is usually done through
:ref:`Parquet <parquet>`, :ref:`IPC format <ipc>` (:ref:`feather`), :ref:`CSV <csv>` or
:ref:`Line-Delimited JSON <json>` formats.

Performing Computations
-----------------------

Arrow ships with a bunch of compute functions that can be applied
to its arrays and tables, so through the compute functions 
it's possible to apply transformations to the data

.. ipython:: python

    import pyarrow.compute as pc

    pc.value_counts(birthdays_table["years"])

See :ref:`compute` for a list of available compute functions and
how to use them.

Working with large data
-----------------------

> Arrow also provides the :class:`pyarrow.dataset` api to work with

Reviewer: "API" uppercased?

Author: :+1:

.. regarding copyright ownership.  The ASF licenses this file
.. to you under the Apache License, Version 2.0 (the
.. "License"); you may not use this file except in compliance
.. with the License.  You may obtain a copy of the License at

..   http://www.apache.org/licenses/LICENSE-2.0

.. Unless required by applicable law or agreed to in writing,
.. software distributed under the License is distributed on an
.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
.. KIND, either express or implied.  See the License for the
.. specific language governing permissions and limitations
.. under the License.

.. _getstarted:

Getting Started
===============

Arrow manages data in Arrays (:class:`pyarrow.Array`), which can be
grouped in tables (:class:`pyarrow.Table`) to represent columns of data
in tabular data.

Arrow also exposes supports for various formats to get those tabular
data in and out of disk and networks. Most commonly used formats are
Parquet (:ref:`parquet`) and the IPC format (:ref:`ipc`). 

Creating Arrays and Tables
--------------------------

Arrays in Arrow are collections of data of uniform type. That allows

> arrow to use the best performing implementation to store the data and

Reviewer 2: ```suggestion
Arrow to use the best performing implementation to store the data and
```

maybe "optimal implementation" instead of "best performing"?

> perform computation of it. So each array is meant to have data and

Reviewer 2: ```suggestion
perform computations on it. So each array is meant to have data and
```

a type

.. ipython:: python

    import pyarrow as pa

    days = pa.array([1, 12, 17, 23, 28], type=pa.int8())

> multiple arrays can be combined in tables to form the columns

Reviewer 2: ```suggestion
Multiple arrays can be combined in tables to form the columns
```

in tabular data according to a provided schema

.. ipython:: python

    months = pa.array([1, 3, 5, 7, 1], type=pa.int8())
    years = pa.array([1990, 2000, 1995, 2000, 1995], type=pa.int16())

    birthdays_table = pa.table([days, months, years], 
                               schema=pa.schema([
                                    ('days', days.type),
                                    ('months', months.type),

>                                     ('years', years.type)

Author: Fair point, I wrote that tables are arrays+schema but in reality they are arrays+names as arrays already provide type information.

                               ]))
    
    birthdays_table

See :ref:`data` for more details.

Saving and Loading Tables
-------------------------

Once you have a tabular data, Arrow provides out of the box
the features to save and restore that data for common formats
like parquet

.. ipython:: python   

    import pyarrow.parquet as pq

    pq.write_table(birthdays_table, 'birthdays.parquet')

> Once you have your data on disk, loading it back is as easy,

Reviewer 2: In general it's good to avoid terms as "easy" and "just" in explanations, according to best practices (not that I always do that!)

Author: You are correct, I wanted to express that it's just 1 line of code like writing, but I couldn't come with a better phrase than "as easy". I tried rewording it now.

and Arrow is heavily optimized for memory and speed so loading
data will be as quick as possible

.. ipython:: python

    reloaded_birthdays = pq.read_table('birthdays.parquet')

    reloaded_birthdays

Saving and loading back data in arrow is usually done through

> :ref:`parquet`, :ref:`ipc`, :ref:`csv` or :ref:`json` formats.

Author: :+1: replaced with explicit names

Performing Computations
-----------------------

Arrow ships with a bunch of compute functions that can be applied

> to its arrays, so through the compute functions it's possible to apply

Author: :+1: added tables

transformations to the data

.. ipython:: python

    import pyarrow.compute as pc

    pc.value_counts(birthdays_table["years"])

> See :ref:`compute` for a list of available compute functions and

Reviewer 2: This links to the python page, which doesn't actually have a list of them ... (but not sure if directly linking to the C++ ones is better, though, it's just not ideal ;))

Author: Yes, it's actually something I want to fix as we already have https://arrow.apache.org/docs/python/api/compute.html which does list compute functions in python

how to use them.

> Working with big data

Reviewer 2: I would personally avoid the term "Big Data" (although you maybe didn't mean "Big Data" with capitals here :), it's still what many people will read)

Not directly sure what would be a good alternative. "Working with datasets" is probably not specific enough. "Working with partitioned datasets"? (or "Working with larger, partitioned datasets" ..)

Author: I see what you mean, but I explicitly avoided using "partitioned data" because I perceive partitioning as the solution to a problem, not the problem itself. And here I was looking for a title that described what the following paragraphs are going to explain how to solve.

I guess we can use "large data"? instead of "big data"? That doesn't change much from the title meaning point of view but it would avoid using a word that people explicitly map to something specific.

Reviewer 2: (personally I would be in favor of using "large" instead of "big", that seems a bit more neutral)

Author: :+1: renamed "big" to "large"

---------------------

Arrow also provides the :class:`pyarrow.dataset` api to work with
big data, which will handle for you partitioning of your data in
smaller chunks

.. ipython:: python

    import pyarrow.dataset as ds

    ds.write_dataset(birthdays_table, "savedir", format="parquet", 
                     partitioning=ds.partitioning(
                        pa.schema([birthdays_table.schema.field("years")])
                    ))

Loading back the partitioned dataset will detect the chunks

.. ipython:: python

>     birthdays_dataset = ds.dataset("savedir", schema=birthdays_table.schema,
>                                    partitioning=ds.partitioning(field_names=["years"]))

Reviewer 2: ```suggestion
    birthdays_dataset = ds.dataset("savedir", format="parquet", partitioning=["years"])
```

I would explicitly pass the format, because we don't actually "infer" it (this just happens to work because "parquet" is the default format, but it won't work if used a different format for writing). 
The schema is not really needed in this case, I think.

    birthdays_dataset.files

and will lazily load chunks of data only when iterating over them

.. ipython:: python

    import datetime

    current_year = datetime.datetime.utcnow().year
    for table_chunk in birthdays_dataset.to_batches():

>         print("AGES", pc.abs(pc.subtract(table_chunk["years"], current_year)))

Reviewer 2: ```suggestion
        print("AGES", pc.subtract(current_year, table_chunk["years"]))
```

(didn't test!)

For further details on how to work with big datasets, how to filter them,

> how to project them etc... refer to :ref:`dataset` documentation.

Reviewer 2: ```suggestion
how to project them, etc., refer to :ref:`dataset` documentation.
```
